meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,484 @@
+const PIECE_GROUP_INFIX: string = '_piece_group_'
+
+// These priority values are the same values used by Core
+const HIGH_PRIORITY: number = 5
+const MEDIUM_PRIORITY: number = 1
+const BASELINE_PRIORITY: number = 0
+const LOW_PRIORITY: number = -1
+
+export class SuperflyTimelineBuilder implements TimelineBuilder {
+	constructor(private objectCloner: ObjectCloner) {}
+
+	public getBaseTimeline(): Timeline {
+		return { timelineGroups: [] }
+	}
+
+	public buildTimeline(rundown: Rundown): Timeline {
+		const timeline: Timeline = {
+			timelineGroups: [],
+		}
+
+		// TODO: Find LookAhead
+
+		const baselineGroup: TimelineObjectGroup = this.createBaselineGroup(rundown)
+		timeline.timelineGroups.push(baselineGroup)
+
+		const activePartTimelineGroup: ActivePartTimelineObjectGroup =
+			this.createTimelineObjectGroupForActivePart(rundown)
+		timeline.timelineGroups.push(activePartTimelineGroup)
+
+		const previousPart: Part | undefined = rundown.getPreviousPart()
+		if (previousPart) {
+			const previousPartTimelineGroup: TimelineObjectGroup = this.createTimelineObjectGroupForPreviousPart(
+				previousPart,
+				rundown.getActivePart(),
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(previousPartTimelineGroup)
+		}
+
+		const infinitePiecesTimelineGroups: TimelineObjectGroup[] =
+			this.createTimelineObjectGroupsForInfinitePieces(rundown)
+		timeline.timelineGroups.push(...infinitePiecesTimelineGroups)
+
+		if (activePartTimelineGroup.shouldAutoNextAtPointInTime > 0) {
+			const nextPartTimeline: TimelineObjectGroup = this.createTimelineForNextPart(
+				rundown,
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(nextPartTimeline)
+
+			if (Number.isNaN(activePartTimelineGroup.enable.duration)) {
+				throw new UnsupportedOperation(
+					`Duration of activePartTimelineGroup must be a number! Got: ${activePartTimelineGroup.enable.duration}`
+				)
+			}
+			timeline.autoNext = {
+				epochTimeToTakeNext:
+					activePartTimelineGroup.shouldAutoNextAtPointInTime -
+					rundown.getNextPart().getTimings().previousPartContinueIntoPartDuration,
+			}
+		}
+
+		// TODO: Call Blueprint "onTimelineGenerate". This will most likely need some tweaks.
+
+		return timeline
+	}
+
+	private createBaselineGroup(rundown: Rundown) {
+		const baselineGroup: TimelineObjectGroup = {
+			id: BASELINE_GROUP_ID,
+			isGroup: true,
+			children: rundown.getBaseline(),
+			enable: {
+				while: '1',
+			},
+			priority: BASELINE_PRIORITY,
+			layer: '',
+			content: {
+				type: TimelineObjectType.GROUP,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+		return baselineGroup
+	}
+
+	private createTimelineObjectGroupForActivePart(rundown: Rundown): ActivePartTimelineObjectGroup {
+		const now: number = Date.now()
+
+		const activePart: Part = rundown.getActivePart()
+
+		const currentPartEnable: TimelineEnable = {
+			start: now,
+		}
+
+		let shouldAutoNextAtPointInTime: number = 0
```

Comment:
Minor thing: `shouldAutoNextAtPointInTime` sounds to me like it would be a boolean saying it should autoNext at some point in time. Rather than a timestamp where we should do the autoNext.

Reply:
Do you have a suggestion that would make it more clear?

Reply:
I pondered a bit but i found it difficult. But maybe something like `autoNextTime` or `autoNextCountdown?`

Reply:
We could take the same approach as in the CallbackScheduler and call it `autoNextEpochTime`.

Reply:
Updated